### PR TITLE
Gyroscope-driven animations of empty chats

### PR DESCRIPTION
**Pull request by wojtus7**
### Details

Add gyroscope-driven animation of the background image in empty chat.

### Fixed Issues

$ https://github.com/Expensify/App/issues/16366

### Tests

Note: This is feature that is only affecting physical devices because it actively using gyroscope. Running on simulator/emulator and web shouldn't be affected by those changes.
 
1. Open app
2. Go into empty conversation
3. Move device around

- [ ] Verify that no errors appear in the JS console

### QA Steps

1. Take physical android or iOS phone
2 Open app
3. Go into empty conversation
4. Move device around

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [X] I linked the correct issue in the `### Fixed Issues` section above
- [X] I wrote clear testing steps that cover the changes made in this PR
    - [X] I added steps for local testing in the `Tests` section
    - [X] I added steps for the expected offline behavior in the `Offline steps` section
    - [X] I added steps for Staging and/or Production testing in the `QA steps` section
    - [X] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [X] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I ran the tests on **all platforms** & verified they passed on:
    - [X] Android / native
    - [X] Android / Chrome
    - [X] iOS / native
    - [X] iOS / Safari
    - [X] MacOS / Chrome / Safari
    - [X] MacOS / Desktop
- [X] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [X] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [X] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [X] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] I verified that if a function's arguments changed that all usages have also been updated correctly
- [X] If a new component is created I verified that:
    - [X] A similar component doesn't exist in the codebase
    - [X] All props are defined accurately and each prop has a `/** comment above it */`
    - [X] The file is named correctly
    - [X] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [X] The only data being stored in the state is data necessary for rendering and nothing else
    - [X] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [X] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [X] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [X] All JSX used for rendering exists in the render method
    - [X] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If a new CSS style is added I verified that:
    - [X] A similar style doesn't already exist
    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [X] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [X] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

image

</details>

<details>
<summary>Mobile Web - Chrome</summary>

IMG_8721

</details>

<details>
<summary>Mobile Web - Safari</summary>

Zrzut ekranu 2023-07-27 o 17 10 21

</details>

<details>
<summary>Desktop</summary>

image

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/42337257/da5e0e11-2bba-4f28-9daf-64f6d0a41634

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/42337257/c504ecf5-d4f7-42eb-a688-d85df286f1f2

</details>

### CODE REVIEW

**Comment by mollfpr** [comment]
@wojtus7 Could you merge `main`? We have reanimated `3.4.0` and merged https://github.com/Expensify/App/pull/23854 with the fix.

**Comment by mollfpr** [comment]
@wojtus7 Is this ready to test? If yes, @roryabraham can add a `Ready to build` label for testing on the actual device.

**Comment by wojtus7** [comment]
@mollfpr Yup! I think it's ready

**Comment by roryabraham** [comment]
Running a test build [here](https://github.com/Expensify/App/actions/runs/5791369573)

File: src/pages/home/report/ReportActionItemCreated.js
Comment (roryabraham) [inline src/pages/home/report/ReportActionItemCreated.js:56]: NABish move these variables outside the function so they aren't redeclared unnecessarily.

File: src/pages/home/report/ReportActionItemCreated.js
Comment (roryabraham) [inline src/pages/home/report/ReportActionItemCreated.js:61]:
```suggestion
    const xOffset = useSharedValue(0);
```

File: src/pages/home/report/ReportActionItemCreated.js
Comment (roryabraham) [inline src/pages/home/report/ReportActionItemCreated.js:62]:
```suggestion
    const yOffset = useSharedValue(0);
```

File: src/pages/home/report/ReportActionItemCreated.js
Comment (roryabraham) [inline src/pages/home/report/ReportActionItemCreated.js:68]: This comment looks wrong since we're using `x` and `y` now...

File: src/pages/home/report/ReportActionItemCreated.js
Comment (roryabraham) [inline src/pages/home/report/ReportActionItemCreated.js:63]: Does this need to be a shared value? It never changes so it seems like it could just be a constant?

File: src/pages/home/report/ReportActionItemCreated.js
Comment (roryabraham) [inline src/pages/home/report/ReportActionItemCreated.js:74]: NABish but I think this would be a bit clearer:

1. Create a function in `NumberUtils` called `clamp`

    ```js
    /**
     * Clamp a number in a range.
     *
     * @param {Number} num
     * @param {Number} min
     * @param {Number} max
     * @returns {Number}
     */
    function clamp(num, min, max) {
        return Math.min(Math.max(num, min), max);
    }
    ```
    
1. (optional) Replace any uses of `lodashClamp` with `NumberUtils.clamp` (we're trying to move away from `lodash` and `underscore`).
1. Then use it here too:

    ```js
    xOffset.value = NumberUtils.clamp(xOffset.value - y * ANIMATION_BOOST, -IMAGE_OFFSET_Y, IMAGE_OFFSET_Y)
    ```

File: src/pages/home/report/ReportActionItemCreated.js
Comment (roryabraham) [inline src/pages/home/report/ReportActionItemCreated.js:76]: Let's flip this around and return `{}` early if `!isSmallScreenWidth`:

```js
const animatedStyles = useAnimatedStyle(() => {
    if (!isSmallScreenWidth) {
        return {};
    }
    const {x, y} = animatedSensor.sensor.value;
    ...
    ...
});
```

This conforms better to our `prefer-early-return` JS style and also helps prevent any unnecessary computation (even if it does take a negligible amount of time)

File: src/pages/home/report/ReportActionItemCreated.js
Comment (roryabraham) [reply src/pages/home/report/ReportActionItemCreated.js:63]: And if so it can probably be removed and we can just reference `-IMAGE_OFFSET_X` directly?

**Review by roryabraham — CHANGES_REQUESTED** [review]

File: src/pages/home/report/ReportActionItemCreated.js
Comment (wojtus7) [reply src/pages/home/report/ReportActionItemCreated.js]: One thing about it is that code inside useAnimatedStyle is run on UI thread so clamp function would need to be a worklet. To avoid confusion in such scenario I usually create function called something like `clampWorklet` so it's hard to import it without knowing it will break the app if run from JS thread.

File: ios/Podfile.lock
Comment (roryabraham) [inline ios/Podfile.lock:1237]: Can you make sure to use `bundle exec pod install`? We are trying to use the Gemfile to standardize the pod version ([context](https://github.com/Expensify/App/issues/21318)).

**Review by roryabraham — CHANGES_REQUESTED** [review]

File: ios/Podfile.lock
Comment (wojtus7) [reply ios/Podfile.lock]: Ough, sorry 😬

**Review by roryabraham — APPROVED** [review]
Code LGTM, will let @mollfpr test and complete the checklist

**Comment by mollfpr** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android / native
    - [ ] Android / Chrome
    - [ ] iOS / native
    - [ ] iOS / Safari
    - [ ] MacOS / Chrome / Safari
    - [ ] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>



</details>

<details>
<summary>Mobile Web - Chrome</summary>



</details>

<details>
<summary>Mobile Web - Safari</summary>



</details>

<details>
<summary>Desktop</summary>



</details>

<details>
<summary>iOS</summary>



</details>

<details>
<summary>Android</summary>



</details>

**Review by mollfpr — APPROVED** [review]
LGTM and tests well 🚀

File: src/pages/home/report/ReportActionItemCreated.js
Comment (roryabraham) [inline src/pages/home/report/ReportActionItemCreated.js:54]: After testing, this feels overly-sensitive to me. Let's just get rid of this and not boost the animation at all.

File: src/pages/home/report/ReportActionItemCreated.js
Comment (roryabraham) [inline src/pages/home/report/ReportActionItemCreated.js:73]: I think we should switch this:

```suggestion
        xOffset.value = NumberUtils.clampWorklet(xOffset.value + y, -IMAGE_OFFSET_X, IMAGE_OFFSET_X);
```

Such that if you're holding the phone flat and tilt it to the left, the animation shifts to the right. This creates a better perspective effect imo and is more inline with how it behaves in the Y direction

**Review by roryabraham — CHANGES_REQUESTED** [review]
Sorry, requesting a few changes after actually testing this.

**Review by roryabraham — CHANGES_REQUESTED** [review]
Also, this same image is applied in a few places:

- `ReportActionItemCreated`
- `MoneyReportView`
- `MoneyRequestView`

Can it a generic `AnimatedEmptyStateBackground` component and use that same thing in all three locations please?

File: src/pages/home/report/AnimatedEmptyStateBackground.js
Comment (roryabraham) [inline src/pages/home/report/AnimatedEmptyStateBackground.js:11]: This prop is not needed since you're using `useWindowDimensions` already. Let's remove it.

File: src/pages/home/report/AnimatedEmptyStateBackground.js
Comment (roryabraham) [inline src/pages/home/report/AnimatedEmptyStateBackground.js:35]: NAB but let's make this early return the first line of the `useAnimatedStyle` callback so we don't call the worklets unnecessarily

**Review by roryabraham — CHANGES_REQUESTED** [review]

**Review by roryabraham — CHANGES_REQUESTED** [review]
Testing on:

- iOS native ✅ 
- iOS Chrome ⚪ – no effect, no bugs
- iOS Safari ⚪ – no effect, no bugs
- macOS Chrome ⚪ – no effect, no bugs
- macOS desktop ⚪ – no effect, no bugs
- Android Chrome ⚠️ – works ok, but each tilt direction is inverted from iOS native
- Android Native ⚠️ – same as Android Chrome

Should we be using pitch and yaw instead of x and y so that it's standardized across the iOS and Android coordinate systems?

**Comment by wojtus7** [comment]
Ough 😬
You are right :(

As far as I see only the code for Rotation was adjusted to work same on iOS and Android and different sensors were left as is 😭 
But even in this scenario I think we should stick to the gyroscope `x` and `y` because when using yaw and pitch at certain rotations you can make the image behave really strange and I think gyroscope makes it more robust and easier to blend-in
In the video I move the phone really gently and change in axises make the background jump 😞

https://github.com/Expensify/App/assets/42337257/ca98e06e-102b-46c5-bd1b-a441627bf2ac

I don't know if we should aim to make fix for reanimated or create a patch but returning values [here](https://github.com/software-mansion/react-native-reanimated/blob/f55e831b1e45566f3f28707344e983ade62021e6/android/src/main/java/com/swmansion/reanimated/sensor/ReanimatedSensorListener.java#L73C9-L73C9) not negative makes iOS and android behave the same

**Comment by roryabraham** [comment]
@wojtus7 we can use a patch as long as we also create an upstream PR and include a link to it in this PR description.

**Comment by wojtus7** [comment]
@roryabraham Just to keep you posted PR fixing Android (And adding reasonable examples for reanimated so it will be easier to keep it working in future) is currently being reviewed [here](https://github.com/software-mansion/react-native-reanimated/pull/4933)

**Comment by wojtus7** [comment]
Fixes for [Android](https://github.com/software-mansion/react-native-reanimated/pull/4933) as well as [Web Android ](https://github.com/software-mansion/react-native-reanimated/pull/4956) were merged and will be released in Reanimated 3.5.0 in about ~week. 

@roryabraham Do you think we should add patch or wait for 3.5.0 and bump reanimated then?

**Comment by roryabraham** [comment]
@wojtus7 this is a non-urgent feature so let's wait for the release, then upgrade Reanimated in this PR. Thanks!

**Comment by wojtus7** [comment]
@roryabraham 3.5.0 is out and I prepared a PR [here](https://github.com/Expensify/App/pull/27193) so we can test it and merge 🎉

**Comment by roryabraham** [comment]
bumped C+ for review of linked PR

**Comment by mollfpr** [comment]
Testing it now.

**Comment by wojtus7** [comment]
I checked it out, and it looks like with reanimated 3.5.4 everything works well 👍

**Comment by roryabraham** [comment]
Ok, I was seeing these same tests failing on main earlier today, so I don't think they're related. Going to create an AdHoc build to make this easier to test

**Comment by roryabraham** [comment]
I think the failing test here is unrelated – was failing on main for me last week. Going to retry in case it's just flaky, but you may need to merge main to get it fixed

**Comment by roryabraham** [comment]
Retested on all platforms:

- iOS native ✅ 
- iOS Chrome ⚪ – no effect, no bugs
- iOS Safari ⚪ – no effect, no bugs
- macOS Chrome ⚪ – no effect, no bugs
- macOS desktop ⚪ – no effect, no bugs
- Android Chrome ✅ 
- Android Native ✅

**Review by roryabraham — APPROVED** [review]

**Comment by roryabraham** [comment]
@wojtus7 Please merge main into this branch to make tests pass – it's the last thing that's needed before I can merge this.

**Comment by wojtus7** [comment]
@roryabraham Done 🫡

**Pull request merged**
